src/ReinforcementLearningCore/src/core/run.jl

```diff
@@ -83,39 +83,41 @@ function _run(policy::AbstractPolicy,
         stop_condition::AbstractStopCondition,
         hook::AbstractHook,
         reset_condition::AbstractResetCondition)
-    hook(PreExperimentStage(), policy, env)
-    policy(PreExperimentStage(), env)
+    update!(hook, PreExperimentStage(), policy, env)
+    update!(policy, PreExperimentStage(), env)
     is_stop = false
     while !is_stop
         reset!(env)
-        policy(PreEpisodeStage(), env)
-        hook(PreEpisodeStage(), policy, env)
+        update!(policy, PreEpisodeStage(), env)
+        update!(hook, PreEpisodeStage(), policy, env)
 
         while !reset_condition(policy, env) # one episode
-            policy(PreActStage(), env)
-            hook(PreActStage(), policy, env)
+            update!(policy, PreActStage(), env)
+            update!(hook, PreActStage(), policy, env)
+
+            action = RLBase.plan!(policy, env)
+            act!(env, action)
 
-            env |> policy |> env
             optimise!(policy)
 
-            policy(PostActStage(), env)
-            hook(PostActStage(), policy, env)
+            update!(policy, PostActStage(), env)
+            update!(hook, PostActStage(), policy, env)
 
-            if stop_condition(policy, env)
+            if stop(stop_condition, policy, env)
                 is_stop = true
-                policy(PreActStage(), env)
-                hook(PreActStage(), policy, env)
-                policy(env)  # let the policy see the last observation
+                update!(policy, PreActStage(), env)
+                update!(hook, PreActStage(), policy, env)
+                RLBase.plan!(policy, env)  # let the policy see the last observation
                 break
             end
         end # end of an episode
 
         if is_terminated(env)
-            policy(PostEpisodeStage(), env)  # let the policy see the last observation
-            hook(PostEpisodeStage(), policy, env)
+            update!(policy, PostEpisodeStage(), env)  # let the policy see the last observation
+            update!(hook, PostEpisodeStage(), policy, env)
```

Comment:
Shouldn't we add a `plan!` line here to see the last action as well ?

Reply:
👍 Ok to save this for another PR?